### PR TITLE
gcc12: 12.1.0 -> 12.2.0

### PR DESCRIPTION
**Pull request by trofi**
Dropped upstreamed musl patches.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by trofi** [comment]
> * [ ]  check conflicts with [gcc12: apply working patch for darwin-aarch64 #186469](https://github.com/NixOS/nixpkgs/pull/186469)

I think it does:

```diff
--- a/pkgs/development/compilers/gcc/12/default.nix
+++ b/pkgs/development/compilers/gcc/12/default.nix
@@ -66,7 +66,7 @@ let majorVersion = "12";
         ../gnat-cflags-11.patch
         ../gcc-12-gfortran-driving.patch
         ../ppc-musl.patch
-      ] ++ optional (stdenv.isDarwin && stdenv.isAarch64) (fetchpatch {
+      ] ++ optional true (fetchpatch {
         url = "https://github.com/Homebrew/formula-patches/raw/76677f2b/gcc/gcc-12.1.0-arm.diff";
         sha256 = "sha256-IcCYHSCAElJqTAZELJnRRWo0/SlkgQzSvoMjYr3pgD0=";
       })
```

```
$ nix build --no-link -f. gcc12 -L
...
gcc> applying patch /nix/store/3qqvx0c72pav14n6f7niw7ppbapj8nk1-gcc-12.1.0-arm.diff
...
gcc> 1 out of 6 hunks FAILED -- saving rejects to file gcc/config/aarch64/aarch64-builtins.cc.rej
gcc> 3 out of 11 hunks FAILED -- saving rejects to file gcc/config/darwin.h.rej

```

**Comment by vcunat** [comment]
They don't have an updated patch yet: https://github.com/Homebrew/formula-patches/tree/master/gcc

**Comment by tjni** [comment]
From my point of view, feel free to do whatever you want to get this upgrade through, even if you have to break this on aarch64-darwin. I can take a look again afterward to see if it's fixable.

**Comment by trofi** [comment]
I commented out outdated patch.

**Comment by vcunat** [comment]
As a simple option, I can imagine making the version (and hash) conditional, keeping 12.1.0 for aarch64-darwin (for now at least).  The musl patches surely don't matter there.

**Comment by tjni** [comment]
Please use the following patch. I cherry-picked commits from https://github.com/iains/gcc-12-branch over to gcc-12.2.

https://github.com/tjni/gcc-12-branch/compare/releases/gcc-12.2...gcc-12-2-darwin.diff

**Comment by trofi** [comment]
> You can use the following patch. I cherry-picked commits from https://github.com/iains/gcc-12-branch over to gcc-12.2.
> 
> https://github.com/tjni/gcc-12-branch/compare/releases/gcc-12.2...gcc-12-2-darwin.diff
> 
> Can replace with officially maintained one once it is created.

Nice! Thank you! Used your patch link instead and added a `TODO` comment in the code.

File: pkgs/development/compilers/gcc/12/default.nix [hunk 1]
```diff
@@ -66,12 +66,11 @@ let majorVersion = "12";
...
         ../ppc-musl.patch
-        # Backports. Should be removed with 12.2.0 release:
-        ./PR106102-musl-poison-cpp.patch
-        ./PR106102-musl-poison-jit.patch
       ] ++ optional (stdenv.isDarwin && stdenv.isAarch64) (fetchpatch {
-        url = "https://github.com/Homebrew/formula-patches/raw/76677f2b/gcc/gcc-12.1.0-arm.diff";
-        sha256 = "sha256-IcCYHSCAElJqTAZELJnRRWo0/SlkgQzSvoMjYr3pgD0=";
+        # TODO: switch back to Homebrew patches:
+        #   was "https://github.com/Homebrew/formula-patches/raw/76677f2b/gcc/gcc-12.1.0-arm.diff"
+        url = "https://github.com/tjni/gcc-12-branch/compare/releases/gcc-12.2...gcc-12-2-darwin.diff";
```
Comment (risicle) [inline pkgs/development/compilers/gcc/12/default.nix:72]:
```suggestion
        name = "gcc-12-darwin-aarch64.patch";
        url = "https://github.com/tjni/gcc-12-branch/compare/releases/gcc-12.2...gcc-12-2-darwin.diff";
```

Or some other helpful name you think is better.

Comment (trofi) [reply pkgs/development/compilers/gcc/12/default.nix:72]: Good point! Applying.

**Comment by risicle** [comment]
`nixpkgs-review` successful, nixos x86_64 & macos 10.15. `pkgsMusl.gcc12` builds on nixos x86_64.

**Pull request merged**